src/app/icd/ICDCheckInSender.cpp

```diff
@@ -0,0 +1,152 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "ICDCheckInSender.h"
+
+#include <app/InteractionModelEngine.h>
+#include <system/SystemPacketBuffer.h>
+
+#include <protocols/secure_channel/CheckinMessage.h>
+
+#include <lib/dnssd/Resolver.h>
+
+namespace chip {
+namespace app {
+
+using namespace Protocols::SecureChannel;
+
+ICDCheckInSender::ICDCheckInSender(PersistentStorageDelegate * storage, FabricTable * fabricTable,
+                                   Crypto::SymmetricKeystore * symmetricKeyStore)
+{
+    VerifyOrDie(storage != nullptr);
+    VerifyOrDie(fabricTable != nullptr);
+    VerifyOrDie(symmetricKeyStore != nullptr);
+
+    mStorage           = storage;
+    mFabricTable       = fabricTable;
+    mSymmetricKeystore = symmetricKeyStore;
+
+    mAddressLookupHandle.SetListener(this);
+}
+
+ICDCheckInSender::ICDCheckInSender()
+{
+    mAddressLookupHandle.SetListener(this);
+}
+
+void ICDCheckInSender::Init(PersistentStorageDelegate * storage, FabricTable * fabricTable,
+                            Crypto::SymmetricKeystore * symmetricKeyStore)
+{
+    VerifyOrDie(storage != nullptr);
+    VerifyOrDie(fabricTable != nullptr);
+    VerifyOrDie(symmetricKeyStore != nullptr);
+
+    mStorage           = storage;
+    mFabricTable       = fabricTable;
+    mSymmetricKeystore = symmetricKeyStore;
+}
+
+void ICDCheckInSender::OnNodeAddressResolved(const PeerId & peerId, const AddressResolve::ResolveResult & result)
+{
+    mResolveInProgress = false;
+
+    ICDMonitoringEntry entry(mSymmetricKeystore);
+
+    const FabricInfo * f = mFabricTable->FindFabricWithCompressedId(peerId.GetCompressedFabricId());
```

Comment:
Compressed IDs are not guaranteed to be unique.  Ideally we would not be using compressed IDs here, and would instead track which fabric index we are working with to start with.

Reply:
Considering where this function is called, we don't have access to the FabricIndex anymore and there is no easy way of retrieving it either. 

The only two solutions I see for your comments are either : 

- Modifying the `peerID` class to store the FabricIndex
- Keeping the Current MonitoringEntry inside the ICDCheckInSender (that will bypass the need of re-finding the monitoring entry inside the MonitoringTable.

I'll implement the second one since it seems a fair tradeof between RAM usage and code efficiency.